File: 백준/Gold/11729. 하노이 탑 이동 순서/하노이 탑 이동 순서.py


def hanoi(n, start, end, via):
    if n == 1:
        print(start, end)
        return
    hanoi(n - 1, start, via, end)  # 제일 큰거 제외 나머지 중간으로 옮기기
    print(start, end)  # 제일 큰거 3번으로
    hanoi(n - 1, via, end, start)  # 중간으로 옮긴 나머지를 3번으로


n = int(input())
print(2 ** n - 1)
hanoi(n, 1, 3, 2)

File: 백준/Gold/11729. 하노이 탑 이동 순서/하노이 탑 이동 순서.js
const input = require('fs')
  .readFileSync('/dev/stdin', 'utf-8')
  .trim()
  .split('\n');

function hanolTower(n, start, end, result) {
  if (n === 1) {
    result.push(`${start} ${end}`);
    return;
  }

  const other = 6 - start - end;
  hanolTower(n - 1, start, other, result);
  result.push(`${start} ${end}`);
  hanolTower(n - 1, other, end, result);
}

const n = Number(input[0]);
const result = [];

console.log(2 ** n - 1);
hanolTower(n, 1, 3, result);
console.log(result.join('\n'));
